File: 03-searching-and-sorting/03_quick_sort/quick_sort.py
"""
Problem: Quick Sort

Time Complexity: O(N log N) Average
Space Complexity: O(log N) recursion stack
"""

from typing import List

def quick_sort(arr: List[int]) -> List[int]:
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quick_sort(left) + middle + quick_sort(right)

if __name__ == "__main__":
    print("[Python] Quick Sort Test")
    arr = [10, 7, 8, 9, 1, 5]
    sorted_arr = quick_sort(arr)
    print("Original:", arr)
    print("Sorted:  ", sorted_arr)
    assert sorted_arr == sorted(arr)
    print("Quick Sort test passed!")

File: 03-searching-and-sorting/03_quick_sort/quick_sort.cpp
/**
 * Problem: Quick Sort
 * 
 * Time Complexity: O(N log N) Average
 * Space Complexity: O(log N) recursion stack
 */

#include <iostream>
#include <vector>
#include <algorithm>

int partition(std::vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; ++j) {
        if (arr[j] <= pivot) {
            ++i;
            std::swap(arr[i], arr[j]);
        }
    }
    std::swap(arr[i + 1], arr[high]);
    return i + 1;
}

void quickSort(std::vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    std::cout << "[C++] Quick Sort Test\n";
    std::vector<int> arr = {10, 80, 30, 90, 40, 50, 70};
    quickSort(arr, 0, arr.size() - 1);
    for (int val : arr) std::cout << val << " ";
    std::cout << "\n";
    return 0;
}
